File: ios/BuddyPlay/Features/Games/Ludo/LudoViewModel.swift
import Foundation
import Combine
import BuddyCore

@MainActor
final class LudoViewModel: ObservableObject {

    @Published private(set) var state: LudoState
    @Published var lastDieRoll: Int?
    @Published var lastError: String?

    let host: Peer
    let guest: Peer

    init(host: Peer, guest: Peer) {
        self.host = host
        self.guest = guest
        self.state = LudoRules.initialState(host: host, guest: guest)
    }

    /// Roll the die. Random 1..6 for v1; in a 2-device session the host
    /// rolls and broadcasts the result.
    func rollDie() {
        let die = Int.random(in: 1...6)
        lastDieRoll = die

        let player = state.sideToMove
        let legal = LudoRules.legalTokenIndices(in: state, player: player, die: die)
        if legal.isEmpty {
            commit(LudoMove(player: player, diceRoll: die, tokenIndex: nil))
        }
    }

    /// Move the indicated token using the most recent die roll.
    func moveToken(_ tokenIndex: Int) {
        guard let die = lastDieRoll else { return }
        commit(LudoMove(player: state.sideToMove, diceRoll: die, tokenIndex: tokenIndex))
    }

    private func commit(_ move: LudoMove) {
        do {
            let step = try LudoRules.reduce(state, input: move)
            state = step.state
            lastDieRoll = nil
            SfxService.shared.playMove()
            if step.outcome != nil { SfxService.shared.playWin() }
        } catch {
            lastError = "\(error)"
        }
    }

    var legalTokenIndices: [Int] {
        guard let die = lastDieRoll else { return [] }
        return LudoRules.legalTokenIndices(in: state, player: state.sideToMove, die: die)
    }

    var currentTurn: UUID? { LudoRules.currentTurn(state) }
}

File: apps/BuddyPlay/android/app/src/main/java/com/americangroupllc/buddyplay/games/ludo/LudoViewModel.kt
package com.americangroupllc.buddyplay.games.ludo

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import com.americangroupllc.buddyplay.core.domain.GameStateReducer
import com.americangroupllc.buddyplay.core.domain.LudoMove
import com.americangroupllc.buddyplay.core.domain.LudoRules
import com.americangroupllc.buddyplay.core.domain.LudoState
import com.americangroupllc.buddyplay.core.models.Peer
import kotlin.random.Random

class LudoViewModel(host: Peer, guest: Peer) : ViewModel() {
    val host = host
    val guest = guest

    var state: LudoState by mutableStateOf(LudoRules.initialState(host, guest))
        private set
    var lastDieRoll: Int? by mutableStateOf(null)
        private set
    var lastError: String? by mutableStateOf(null)
        private set

    fun rollDie() {
        val die = Random.nextInt(1, 7)
        lastDieRoll = die
        val player = state.sideToMove
        val legal = LudoRules.legalTokenIndices(state, player, die)
        if (legal.isEmpty()) {
            commit(LudoMove(player, die, null))
        }
    }

    fun moveToken(tokenIndex: Int) {
        val die = lastDieRoll ?: return
        commit(LudoMove(state.sideToMove, die, tokenIndex))
    }

    val legalTokenIndices: List<Int>
        get() = lastDieRoll?.let { LudoRules.legalTokenIndices(state, state.sideToMove, it) } ?: emptyList()

    private fun commit(move: LudoMove) {
        try {
            val step = LudoRules.reduce(state, move)
            state = step.state
            lastDieRoll = null
        } catch (e: GameStateReducer.Error) {
            lastError = e.message
        }
    }
}
